Feature/if statement grammar fix

damniko opened the pull request:
Changes:

- Grammar now supports else-if chaining within an if-statement
- A couple of AstBuilderVisitor tests were implemented specifically for if-statements
- In order for the tests to work, all nodes now override the Equals method to check structural equality (rather than reference equality)

damniko committed:
Changed grammar to allow "else if"-chaining

damniko committed:
Updated visitors to support the new ElseIfNode

damniko committed:
Generated working parser for the grammar

Note: Due to a bug with the ANTLR 4.10.1 plugin in Rider, the parser had to be generated using the ANTLR 4.9.2 plugin for IntelliJ.

damniko committed:
AstBuilderVisitor now disregards empty statements in if-statements

damniko committed:
Implemented unit tests for else-if chaining

damniko committed:
Removed all ANTLR generated files

... so we can re-generate them after merging!

damniko committed:
Merge branch 'feature/TypeChecking' into feature/if-statement-grammar-fix

damniko committed:
Merge branch 'feature/TypeChecking' into feature/if-statement-grammar-fix

Sarmisuper commented on STEP/AST/AstBuilderVisitor.cs:
```diff
@@ -656,23 +665,142 @@ public override AstNode VisitFor_iter_opt([NotNull] STEPParser.For_iter_optConte
+        }
+        
+        // Get all else-if-clauses
+        foreach (var elseIf in context.elseifstmt())
+        {
+            node.ElseIfClauses.Add((ElseIfNode) elseIf.Accept(this));
+        }
```
Hvad hvis der ikke er nogle elseif? Maybe need null check

Sarmisuper commented on STEP/AST/AstBuilderVisitor.cs:
```diff
@@ -656,23 +665,142 @@ public override AstNode VisitFor_iter_opt([NotNull] STEPParser.For_iter_optConte
+            {
+                node.ThenClause.Add((BreakNode) NodeFactory.MakeNode(AstNodeType.BreakNode));
+            }
+            else if(stmt.CONTINUE != null)
```
Her skal du nok have () efter BREAK og CONTINUE.

Sarmisuper commented on STEP/AST/AstBuilderVisitor.cs:
```diff
@@ -656,23 +665,142 @@ public override AstNode VisitFor_iter_opt([NotNull] STEPParser.For_iter_optConte
+                }
+                else if (stmt.CONTINUE != null)
+                {
+                    node.ElseClause.Add((ContNode)NodeFactory.MakeNode(AstNodeType.ContNode));
```
Se forrige kommentar

Sarmisuper commented on STEP/AST/AstBuilderVisitor.cs:
```diff
@@ -656,23 +665,142 @@ public override AstNode VisitFor_iter_opt([NotNull] STEPParser.For_iter_optConte
+            {
+                node.Body.Add((BreakNode)NodeFactory.MakeNode(AstNodeType.BreakNode));
+            }
+            else if (stmt.CONTINUE != null)
```
Forrige kommentar

Sarmisuper commented on STEP/AST/Nodes/ArrDclNode.cs:
```diff
@@ -8,4 +8,15 @@ public class ArrDclNode : VarDclNode
+        {
+            return other.Size == Size
+                && other.IsId == IsId
+                && Equals(other.IdRight, IdRight);
```
Her mangler der noget. Der mangles at tjekke på nodernes Right property. Altså den property der bruges, når højre side af declarationen er en array literal og ikke et id

Sarmisuper replied:
Siden de er null hvis de ikke sættes (altså IdRight og Right), så er det nok bare at tilføje endnu en &&

Sarmisuper commented on STEP/AST/Nodes/ForNode.cs:
```diff
@@ -9,4 +9,16 @@ public class ForNode : StmtNode
+            return Equals(other.Initializer, Initializer)
+                && Equals(other.Limit, Limit)
+                && Equals(other.Update, Update)
+                && Body.SequenceEqual(Body);
```
Forkert. Burde være other.Body

Sarmisuper commented on STEP/AST/Nodes/IdNode.cs:
```diff
@@ -8,4 +8,15 @@ public class IdNode : ExprNode
+        if (obj is IdNode other)
+        {
+            return Equals(other.Id, Id)
+                && Equals(other.IsArray, IsArray)
```
IsArray er obsolete. Reminder at den property skal slettes, siden at Type propertien beskriver om det er et array.

Sarmisuper commented on STEP/AST/Nodes/NegNode.cs:
```diff
@@ -4,4 +4,13 @@ public class NegNode : ExprNode {
+    {
+        if (obj is NegNode other)
+        {
+            return Equals(other.Left, Left) && Equals(other.Right, Right);
```
Technically correct, but wrong. På en NegNode sættes Left propertien kun. Right sættes aldrig (siden neg jo er en unary operation). Så slet tjekket om Right er equal

Sarmisuper commented on STEP/AST/Nodes/ParenNode.cs:
```diff
@@ -4,4 +4,13 @@ public class ParenNode : ExprNode {
+    {
+        if (obj is ParenNode other)
+        {
+            return Equals(other.Left, Left) && Equals(other.Right, Right);
```
Samme som ved NegNode.

Sarmisuper commented on STEP/AST/Nodes/UMinusNode.cs:
```diff
@@ -4,4 +4,13 @@ public class UMinusNode : ExprNode {
+    {
+        if (obj is UMinusNode other)
+        {
+            return Equals(other.Left, Left) && Equals(other.Right, Right);
```
Samme som ved NegNode

Sarmisuper commented on STEP/TypeVisitor.cs:
```diff
@@ -478,6 +478,23 @@ public void Visit(AssNode n) { // Add constant check, needs to be in ST (i think
+    public void Visit(ElseIfNode n)
+    {
+        n.Condition.Accept(this);
+        if (n.Type != TypeVal.Boolean)
```
This is wrong, right? Burde det ikke være n.Condition.Type != TypeVal.Boolean

Sarmisuper replied:
Og det burde jo så være .Type.ActualType tror jeg

Sarmisuper commented on STEP/TypeVisitor.cs:
```diff
@@ -478,6 +478,23 @@ public void Visit(AssNode n) { // Add constant check, needs to be in ST (i think
+        n.Condition.Accept(this);
+        if (n.Type != TypeVal.Boolean)
+        {
+            n.Type = TypeVal.Error;
```
Den her er dog ok i think...

Sarmisuper replied:
Men måske igen .Type.ActualType

Sarmisuper commented on StepTests/ASTVisitorTests.cs:
```diff
@@ -26,12 +27,13 @@ private STEPParser.ProgramContext GetParseTree(string input)
+
         var parseTree = GetParseTree(program);
 
         var idNode = new IdNode() { Id = "x" };
```
Her mangles at typen af idnoden sættes

Sarmisuper reviewed (changes requested):
Looks very well done. Er meget fan af din måde at build if-statements i AstBuilder, den måde havde jeg ikke tænkt på, genius! Der er dog nogle ting som skal ændres someplaces, har lagt comments til det.

damniko committed:
Fixes based on Sarmi's review

damniko committed:
Merge remote-tracking branch 'origin/feature/TypeChecking' into feature/if-statement-grammar-fix

damniko committed:
Fixed type-checking for ElseIfNode and other small fixes

- Removed IsArray from IdNode
- Removed redeclaration of pindcl in grammar file

KarmaKamikaze commented on STEP/ANTLR gens/STEP.g4:
```diff
@@ -294,7 +312,7 @@ grammar STEP;
   COMMA                         : ',';
   INTLITERAL                    : DIGIT+;
-  NUMLITERAL                    : DIGIT+ ('.' DIGIT*)?;
+  NUMLITERAL                    : (DIGIT+ ('.' DIGIT*)? | '.' DIGIT+);
```
Do we really want this?

damniko replied:
fixed! nu accepterer den kun tal med komma hvis der er tal på begge sider!

KarmaKamikaze commented on STEP/AST/Nodes/IfNode.cs:
```diff
@@ -6,10 +6,67 @@ public class IfNode : StmtNode
+        if(obj is IfNode other) 
+        {
+
+            //TODO: fix this shit with SequenceEquals
```
👀

KarmaKamikaze reviewed (approved):
Det er fanme en lækker måde at starte mandag morgen på 👍

damniko committed:
Changed numliteral to only accept '.' with numbers on both sides

damniko committed:
Simplified Equals method in IfNode
